### ISSUE TITLE
Column resizing

### ISSUE THREAD

**Issue by jouni** [vaadin/vaadin-grid#145]
The end user should be able to resize the column widths by dragging the column dividers.

Note that work is being done in the Vaadin Framework Grid for this feature, so sync with that team before starting to implement this feature.

Tracking issue for the the Framework (until they move stuff over to GitHub): https://dev.vaadin.com/ticket/16838

Update (Aug 2016): the feature shipped in Vaadin Framework 7.6 some months ago.

**Comment by jounik** [comment]
Graphical design needed.

**Comment by tuomasnironen** [comment]
[Graphical designs](https://drive.google.com/drive/u/0/folders/0BzkWbWvgmkIxTHJQZDY3OXY2SVU) available for review

**Comment by Saulis** [comment]
I'm thinking the header column divider could benefit from a hover highlight colour - just to clearly show which divider the cursor is on top of when there are two dividers very close to each other.

A question from UX perspective (and excluding all possible technical issues): when dragging the header column divider, is the column width also being updated realtime in the rows? Or just in the header and the rows are updated after releasing the header divider?

I'm just thinking about the row column divider line: if there is one, I think it should move in sync with the header column divider line.

**Comment by tuomasnironen** [comment]
* Added alternative designs for column divider when hovering (blue). But I'm not sure would this make any difference if two dividers are very close to each other and only one divider is shown at a time (depending on which column the user is hovering). Alternative approach would be to show a gray divider on _all_ columns when hovering on header, and a blue divider if the cursor hovers on a specific column edge. Then you would see the difference more clearly but it also adds visual load. Therefore I would still vote for the first design and a bit more subtle approach.
* I think you should sync head and body rows when dragging a column edge (all technical issues aside :) ) This is how dragging a column works on most spreadsheets and grids in my experience.

**Comment by Saulis** [comment]
>> any difference if two dividers are very close to each other and only one divider is shown at a time (depending on which column the user is hovering).

Ah, right - sorry - I didn't think about it like that. I thought the header column divider line is always visible to indicate which columns are resizable and which aren't.

**Comment by jounik** [comment]
This feature is already implemented in the GWT version, and we should not
change the behavior here. Only the visual styles should be applied. I don’t
think it is technically possible now to only show the body divider when
resizing, and we need to evaluate if that is something we absolutely want
to add.
ti 26. tammikuuta 2016 klo 12.33 Sauli Tähkäpää <notifications@github.com>
kirjoitti:

**Comment by components-stalker** [comment]
You can try out the Framework implementation at
http://demo.vaadin.com/sampler/#ui/grids-and-trees/grid/features
ti 26. tammikuuta 2016 klo 12.48 Jouni Koivuviita <notifications@github.com>
kirjoitti:

**Comment by ericwoerner** [comment]
Any idea when column resizing will be released?  The grid works great but this is a needed feature for my application.

**Comment by jouni** [comment]
I believe @djbowman1 was working on this currently – or is it just the column reordering you are working on? If not, I think it should be fairly easy to follow along the column reordering patch to provide this feature, as it is also available in the internal GWT implementation, but not exposed to the Polymer element.

**Comment by ericwoerner** [comment]
Awesome!  The vaadin-grid is working great we just need the column resizing before we can use it in our production environment.

Thanks!

**Comment by djbowman1** [comment]
@jouni Im just doing column reordering, however there is talk with other people I work with that column resizing may be in the future, but we're not completely sure

**Comment by jstanley1730** [comment]
@jouni we are looking into starting development on this feature, would you be able to provide a rough estimate on how long it should take, or how complex it is?  And also point us to a starting place?

thanks

**Comment by jouni** [comment]
@jstanley1730, are you still looking for an estimate, or did you already start working on this?

**Comment by jouni** [comment]
Also, I shared the folder for the [graphical designs](https://drive.google.com/drive/u/0/folders/0BzkWbWvgmkIxTHJQZDY3OXY2SVU) to everyone, so you can take a look how it was designed to look like.

**Comment by jstanley1730** [comment]
@jouni we made a rough guess on estimate and started work.  Our engineer ran into a issue and posted question to gitter.  Here is copy:

David Poling @davidpoling Aug 11 14:21
@jouni I've exposed two methods in the JSColumn.java, setResizable and getResizable and also added a column.setResizable to the configureColumnsFromDom method in GridLightDomTable.java. I can then configure a resizable attribute for a JSColumn through javascript, but the columns aren't resizable.
Is there anything in particular I'm missing?

Can you provide some guidance as to what it will take to implement? Thanks

**Comment by tomivirkki** [comment]
Hi @jstanley1730 

Tried reproducing what you described and it works for me.
If you look at the DOM, you should see a `.vaadin-grid-column-resize-handle` element appearing inside the header cell when the `resizable` property is changed. The handle itself still needs to be styled properly (by default it's not visible as `<vaadin-grid>` doesn't define any style rules for it).

See the [Vaadin framework Grid](https://vaadin.com/grid) for examples on how the handle should be positioned and the [graphical designs](https://drive.google.com/drive/u/0/folders/0BzkWbWvgmkIxTHJQZDY3OXY2SVU) document Jouni linked on how it should look like with the `<vaadin-grid>`.

**Comment by jstanley1730** [comment]
@tomivirkki it was probably the non visible part that got us:)  We will start the styling based on the graphical designs.

**Comment by davidpoling** [comment]
@tomivirkki I was able to get the functionality working 100% and the styling for the resize handle to look like the graphical designs, except for the vertical line that separates columns when the resize handle is moved. How were you able to accomplish this?

**Comment by jouni** [comment]
I think that is not easily possible, and not a requirement for this feature. Perhaps by using a pseudo element (::before/after) you could fake it.

**Comment by jouni** [comment]
@davidpoling here’s the way to fake the vertical line with a pseudo element: https://github.com/vaadin/vaadin-grid/pull/421/commits/d0a2e0c9d61255f3ce60ddd3d23c1ded2207153b#diff-ebae5445d8ff48af6cf6bf15c570f715R686

**Comment by jouni** [comment]
This feature was released in 1.2.0. We’ll create a new issue that will cover 2.0.
